azurecompute/src/main/java/org/jclouds/azurecompute/binders/DeploymentParamsToXML.java

```diff
@@ -16,104 +16,188 @@
-                 .e("RoleSize").t(params.size().getText()).up()
-                 .up() //Role
-                 .up() //RoleList
-                 .e("VirtualNetworkName").t(params.virtualNetworkName()).up();
+               OSVirtualHardDiskParam osDiskParam = roleParam.OSVirtualHardDiskParam();
+               XMLBuilder osDiskBuilder = roleBuilder.e("OSVirtualHardDisk");
+               osDiskBuilder
+                     .e("OSVirtualHardDisk");
+               if (osDiskParam.hostCaching() != null) {
+                  osDiskBuilder.e("HostCaching").t(osDiskParam.hostCaching()).up();
+               }
+
+               osDiskBuilder
+                     .e("DiskName").t(osDiskParam.diskName()).up()
+                     .e("DiskLabel").t(osDiskParam.diskLabel()).up();
+               add(osDiskBuilder, "MediaLink", osDiskParam.mediaLink());
+               add(osDiskBuilder, "SourceImageName", osDiskParam.sourceImageName());
+
+               if (osDiskParam.OS() != null) {
+                  osDiskBuilder.e("OS").t(osDiskParam.OS().toString()).up();
+               }
+               add(osDiskBuilder, "ReSizedSizeInGB", osDiskParam.resizedSizeInGB());
+
+               if (osDiskParam.sourceImageName() == null) {
+                  add(osDiskBuilder, "RemoteSourceImageLink", osDiskParam.remoteSourceImageLink());
+               }
+
+               roleBuilder.up() //OSVirtualHardDisk
+                     .e("RoleSize").t(UPPER_UNDERSCORE.to(UPPER_CAMEL, roleParam.roleSize().getText())).up()
+                     .up(); //Role
+
+            } else {
+               OSVirtualHardDiskParam osDiskParam = roleParam.OSVirtualHardDiskParam();
+               // builder.up(); // configsets
+               // builder.e("ResourceExtensionReferences").up();
+               roleBuilder.e("VMImageName").t(vmImage).up();
+               URI mediaLocation = roleParam.mediaLocation();
+               if (mediaLocation != null) {
+                  roleBuilder.e("MediaLocation").t(mediaLocation.toASCIIString()).up();
+               }
+               RoleSize.Type roleSize = roleParam.roleSize();
+               if (roleSize != null) {
+                  roleBuilder.e("RoleSize").t(roleSize.getText()).up();
+               }
+               // add(builder, "AvailabilitySetName", params.availabilitySetName);
+               Boolean agent = roleParam.provisionGuestAgent();
+               if (agent != null) {
+                  roleBuilder.e("ProvisionGuestAgent").t(agent.toString()).up();
+               }
+
+            }
+            roleBuilder.up(); // role
+         }
+         roleListBuilder.up(); // RoleList
+         add(builder, "VirtualNetworkName", params.virtualNetworkName());
+         builder.up();
          // TODO: Undeprecate this method as forcing users to wrap a String in guava's ByteSource is not great.
          return (R) request.toBuilder().payload(builder.asString()).build();
       } catch (Exception e) {
```

Comment:
It might be better to add a meaningful message inside the catch block when propagating an exception. Otherwise it would be difficult for users to understand which functionality did not work when an exception is raised.
Thanks